File: programming/lab9/backend/src/main/kotlin/lab9/backend/domain/Vehicle.kt
package lab9.backend.domain

import kotlinx.serialization.Serializable
import lab9.backend.common.LocalDateSerializer
import java.time.LocalDate

@Serializable
data class Vehicle private constructor(
    val id: VehicleID,
    val name: String,
    val creatorID: User.UserID,
    val coordinates: Coordinates,
    @Serializable(with = LocalDateSerializer::class)
    val creationDate: LocalDate,
    val enginePower: Double,
    val vehicleType: VehicleType,
    val fuelType: FuelType?,
) {
    enum class FuelType {
        GASOLINE,
        ELECTRICITY,
        MANPOWER,
        PLASMA,
        ANTIMATTER
    }

    enum class Field {
        ID,
        NAME,
        X,
        Y,
        CREATION_DATE,
        ENGINE_POWER,
        VEHICLE_TYPE,
        FUEL_TYPE,
        CREATOR_ID
    }

    enum class VehicleType {
        PLANE,
        SUBMARINE,
        BOAT,
        BICYCLE;
    }

    companion object {
        fun withID(
            id: VehicleID,
            name: String,
            creatorID: User.UserID,
            coordinates: Coordinates,
            creationDate: LocalDate,
            enginePower: Double,
            vehicleType: VehicleType,
            fuelType: FuelType?,
        ): Vehicle = Vehicle(id, name, creatorID, coordinates, creationDate, enginePower, vehicleType, fuelType)

        fun withoutID(
            name: String,
            creatorID: User.UserID,
            coordinates: Coordinates,
            creationDate: LocalDate,
            enginePower: Double,
            vehicleType: VehicleType,
            fuelType: FuelType?,
        ): Vehicle =
            Vehicle(VehicleID.NoID, name, creatorID, coordinates, creationDate, enginePower, vehicleType, fuelType)
    }


    @Serializable
    data class Coordinates(val x: Int, val y: Long?)

    @Serializable
    data class VehicleID(val id: Int) {
        companion object {
            val NoID = VehicleID(-1)
        }
    }
}

File: programming/lab9/frontend/src/api/defs/vehicles/Vehicle.ts
export enum VehicleType {
    PLANE = "PLANE",
    BICYCLE = "BICYCLE",
    BOAT = "BOAT",
    SUBMARINE = "SUBMARINE",
}

export const PossibleVehicleTypes: VehicleType[] = [
    VehicleType.PLANE, VehicleType.BICYCLE, VehicleType.BOAT, VehicleType.SUBMARINE
]

export enum FuelType {
    MANPOWER = "MANPOWER",
    GASOLINE = "GASOLINE",
    ELECTRICITY = "ELECTRICITY",
    PLASMA = "PLASMA",
    ANTIMATTER = "ANTIMATTER"
}

export const PossibleFuelTypes: FuelType[] = [
    FuelType.ANTIMATTER, FuelType.GASOLINE, FuelType.ELECTRICITY, FuelType.PLASMA, FuelType.MANPOWER
]
export type Coordinates = {
    x: number
    y: number | null
}

export enum VehicleColumn {
    ID = "ID",
    NAME = "NAME",
    X = "X",
    Y = "Y",
    CREATION_DATE = "CREATION_DATE",
    ENGINE_POWER = "ENGINE_POWER",
    VEHICLE_TYPE = "VEHICLE_TYPE",
    FUEL_TYPE = "FUEL_TYPE",
    CREATOR_ID = "CREATOR_ID"
}

export type Vehicle = {
    id: number
    name: string
    coordinates: Coordinates
    creationDate: Date,
    enginePower: number,
    vehicleType: VehicleType
    fuelType: FuelType | null
    creatorId: number,
    canEdit: boolean,
    canDelete: boolean,
}